Add flake.nix to facilitate building and installation of N64Recomp.

Rucadi opened the pull request:
In this PR I add a flake.nix file that it's super useful to install and test this program in computer that have the nix language installed. https://nixos.org/

This allows the following:

- Open a bash session with N64Recomp and RSPRecomp in PATH
nix shell 'git+https://github.com/Mr-Wiseguy/N64Recomp?submodules=1#N64Recomp

- Install on your nix profile
nix profile install 'git+https://github.com/Mr-Wiseguy/N64Recomp?submodules=1#N64Recomp

- Run directly the app
nix run 'git+https://github.com/Mr-Wiseguy/N64Recomp?submodules=1#N64Recomp
nix run 'git+https://github.com/Mr-Wiseguy/N64Recomp?submodules=1#RSPRecomp

And also, since it's a flake, add it to your static configuration for your system.

bates64 commented on README.md:
```diff
@@ -52,6 +52,12 @@ RSP microcode can also be recompiled with this tool. Currently there is no suppo
+If you are using nixos or have [nix installed](https://nixos.org/)
+This project can be built and run using nix flakes, to spawn a shell with both, N64Recomp and RSPRecomp programs available you can perform the command:
+
+`nix shell 'git+https://github.com/Mr-Wiseguy/N64Recomp?submodules=1#N64Recomp`
```
Would `nix shell github:Mr-Wiseguy/N64Recomp` be OK instead? (Could be totally wrong here, I'm still new to Nix and especially flakes)

Rucadi replied:
Yes and no, 

the problem is that this repository has submodules, when using submodules, you have to specify the submodules=1, and that's not supported with all uri notations, hence you have to use the full url + ?submodules=1.

If we packaged this for nixpkgs, or forced the package to be in an specific commit (instead of the latest at the time) then we could do it your way, since the submodule resolution would be on the src attribute in the mkDerivaton.

PaulGrandperrin commented:
Hey @Rucadi !
I just want to say that I'm trying to build zelda64 on NixOS too ;)

Here's the state of my progress: https://github.com/Mr-Wiseguy/N64Recomp/issues/13#issuecomment-2105966546

Have you been able to build Zelda ?

Rucadi commented:
> Hey @Rucadi !
> I just want to say that I'm trying to build zelda64 on NixOS too ;)
> 
> Here's the state of my progress: https://github.com/Mr-Wiseguy/N64Recomp/issues/13#issuecomment-2105966546
> 
> Have you been able to build Zelda ?
> 

I wanted this merged as a first step since it's a dependency, I have not tried yet, however, tomorrow I'll be able to give it a try, today I'm not at home 😁

Rucadi commented:
> Hey @Rucadi ! I just want to say that I'm trying to build zelda64 on NixOS too ;)
> 
> Here's the state of my progress: [#13 (comment)](https://github.com/Mr-Wiseguy/N64Recomp/issues/13#issuecomment-2105966546)
> 
> Have you been able to build Zelda ?

It should be straightforward if you have an elf file, just something like:


```
{majorasMaskN64? ./zeldamm.n64 ,majorasMaskElf ? ./mm-n64-us.elf}:
let
 pkgs = import <nixpkgs> {};
 
 _n64Recomp = {stdenv, fetchFromGitHub, cmake} : stdenv.mkDerivation 
 {
        pname = "N64Recomp";
        version = "1.0.0";
        src = fetchFromGitHub {
              owner = "Mr-Wiseguy";
              repo = "N64Recomp";
              rev = "dbf0e623c8a1cbb3e73237803a8bcc29f32816d6";
              deepClone = true;
              fetchSubmodules = true;
              hash = "sha256-RUcBd9vB1oFgveHPBjShSlAv+J26/fGCN0/P6atuyo4=";
            };
        nativeBuildInputs = [cmake];

        installPhase = ''
                mkdir -p $out/bin
                cp N64Recomp $out/bin
                cp RSPRecomp $out/bin
            '';
 };

 N64Recomp = pkgs.callPackage _n64Recomp {};

 _zelda64 = {stdenv, fetchFromGitHub, cmake, pkg-config, SDL2, freetype, shaderc, xrandr,  gtk3} : 
 ELF64: stdenv.mkDerivation
        rec {
            pname = "Zelda64Recomp";
            version = "1.0.0";
            src = fetchFromGitHub {
              owner = "Mr-Wiseguy";
              repo = "Zelda64Recomp";
              rev = "v${version}";
              deepClone = true;
              fetchSubmodules = true;
              hash = "sha256-n9wB0OSVQh3HKQgZcfSpxQ0PKJMdI4i+a4mtcYjZRdE=";
            };

            patchPhase = ''
            cp ${ELF64} ./mm.us.rev1.elf
            N64Recomp ./us.rev1.toml
            '';
            nativeBuildInputs = [cmake pkg-config SDL2 freetype shaderc xrandr gtk3 N64Recomp];
            postInstall = ''
                mkdir $out
                cp -R $src $out
            '';
        };
zelda64 =  pkgs.callPackage _zelda64 {};
in
zelda64 majorasMaskElf
```


Then, if you have the correct elf, it will already generate (This is default.nix and called with nix-build).

The problem resides in generating that .elf, I have a mm repository derivation that generates the .elf file, but it seems that I may be missing something because it complains about the generated elf.

once it's clearer how to generate the elf, this default.nix can be used easily with:

nix-build --arg majorasMaskElf <path_to_elf_file>

and with some modifications, hopefully

nix-build --arg majorasMaskN64 <path_to_n64_file> 

and the elf is generated and used transparently.

Mr-Wiseguy commented:
Thanks for the PR! I've never used nix before, so I'll have to spend some time learning it before I feel comfortable merging this. I'll try to find some time in the next week to do so.

Rucadi commented:
> Thanks for the PR! I've never used nix before, so I'll have to spend some time learning it before I feel comfortable merging this. I'll try to find some time in the next week to do so.

Yep! No problem, this is by no ways urgent nor adds "functionality" to the tool.

I modified my original PR, since you use submodules  and the syntax to get nix with flakes+submodules is not ideal, I fixed the commit to use inside the flake.nix, which allows me to specify there the recursive module resolution.

If you have questions about nix, don't hesitate to ask me or the discourse forums!

Arian-D commented:
I get a hash mismatch when trying to run the flake @Rucadi
```
$ nix run 'git+https://github.com/Rucadi/N64Recomp?submodules=1#N64Recomp'
error: hash mismatch in fixed-output derivation '/nix/store/jc8gkaxv6bkk23zvvhf7nmsk229i4wjm-source.drv':
         specified: sha256-TWr6ctsT3VaSRvsTfVBC5sBPCx4b2rvUuuZFUGO6kH4=
            got:    sha256-ocbXNJt5yc+lW9690MCQhAHLZRwsKg5FkdwzVTo2wP4=
error: 1 dependencies of derivation '/nix/store/j4llv6fn8ggwcqgbw4fjhsw79shx3v4b-N64Recomp-1.0.0.drv' failed to build
```

Rucadi commented:
> * nix shell 'git+https://github.com/Mr-Wiseguy/N64Recomp?submodules=1#N64Recomp
> *

You are totally right, I was able to replicate the error, it seems that git is non-hash reproducible sometimes even for nix, I reverted back to just building the latest sources, can do using: nix shell 'git+https://github.com/rucadi/N64Recomp?submodules=1#N64Recomp'
